### PR TITLE
[REVIEW] Fix the use of the device memory resource

### LINKED ISSUE #5067: [BUG] Fix passing of memory resource internally
**Describe the bug**
Some cases of memory allocation in libcudf do not use the memory resource for allocations that are returned to the caller. Some cases use mr for allocations not returned.

Examples:
https://github.com/rapidsai/cudf/blob/8a6ce93f5fdefcd53ce8d981af662228f40c32bf/cpp/src/column/column_factories.cpp#L116
https://github.com/rapidsai/cudf/blob/8a6ce93f5fdefcd53ce8d981af662228f40c32bf/cpp/src/copying/scatter.cu#L321
https://github.com/rapidsai/cudf/blob/8a6ce93f5fdefcd53ce8d981af662228f40c32bf/cpp/src/replace/replace.cu#L398
https://github.com/rapidsai/cudf/blob/8a6ce93f5fdefcd53ce8d981af662228f40c32bf/cpp/src/stream_compaction/drop_duplicates.cu#L197

### PR DESCRIPTION
**Pull request by vuule**
Fixes #5067 

- Use the passed memory resource to allocate all returned objects.
- Use the default memory resource for internal allocations (temporary objects).
- Pass the right memory resource when creating empty` device_buffer`s; no allocation at the time, but buffers hold the `mr` value and can potentially resize later.
- Minor rvalue related clean up, like removal of some `null_mask` copies.

### CODE REVIEW

File: cpp/include/cudf/detail/reduction.cuh [hunk 1]
```diff
@@ -100,7 +100,7 @@ std::unique_ptr<scalar> reduce(InputIterator d_in,
 {
   auto binary_op      = sop.get_binary_op();
   OutputType identity = sop.template get_identity<OutputType>();
-  rmm::device_scalar<OutputType> dev_result{identity, stream, mr};
+  rmm::device_scalar<OutputType> dev_result(identity, stream);
```
Comment (jrhemstad) [inline cpp/include/cudf/detail/reduction.cuh:103]: Why the change from `{}` to `()`? I much prefer the former unless the object has an init-list ctor you're trying to avoid (which device_scalar does not).

File: cpp/include/cudf/detail/gather.cuh [hunk 2]
```diff
@@ -245,7 +245,7 @@ struct column_gatherer_impl<dictionary32, MapItType> {
     auto indices_column = std::make_unique<column>(data_type{INT32},
                                                    static_cast<size_type>(output_count),
                                                    std::move(*(contents.data.release())),
-                                                   rmm::device_buffer{},
+                                                   rmm::device_buffer(0, stream, mr),
```
Comment (jrhemstad) [inline cpp/include/cudf/detail/gather.cuh:248]: This seems unnecessary.

File: cpp/include/cudf/detail/reduction.cuh (hunk 1 above)
Comment (vuule) [reply cpp/include/cudf/detail/reduction.cuh:103]: I vaguely remember @codereport saying that we should prefer `()`, but did not really dig up the convo on Slack to verify. I don't have a preference, may have just misremembered.

File: cpp/include/cudf/detail/gather.cuh (hunk 2 above)
Comment (vuule) [reply cpp/include/cudf/detail/gather.cuh:248]: I'm sure many of my changes from default `device_buffer` ctor to `(0, stream, mr)` are pointless ATM. Some may become relevant if the output changes such that resize is called on the empty buffer (not really possible here). 
Regardless, IMO it does not hurt to call the right ctor here even if it does not affect the allocation. Also, I find it easier to use mr for all returned objects than to consider if an allocation really happens. 
Do you see a downside to this change?

**Comment by vuule** [comment]
> 1. Why the change from `{}` to `()`? Generally we use the former unless you need to use the latter (do to init list ctors).

I got it mixed up. I'll fix it :)

> 2. Change of using the default `device_buffer` ctor to `(0, stream, mr)`. I think I understand why this would be important to do, but curious what inspired you.

I noticed a few days ago that `device_buffer` holds on to the `mr` value and has `resize()`. This means that the passed `mr` can potentially be used even if the initial buffer size is zero and no allocation happens at construction time. Given that it's somewhat easier to use `mr` consistently on all returned buffers (empty or not), and there's no overhead from passing the correct `mr`, I went for it.

File: cpp/include/cudf/detail/reduction.cuh (hunk 1 above)
Comment (jrhemstad) [reply cpp/include/cudf/detail/reduction.cuh:103]: Unfortunately the guidance is a bit confusing. The rule is to prefer `{}` _unless_ the object you are constructing has an initializer list constructor. E.g.,

```
rmm::device_buffer{0,stream,mr}; // OKAY, device_buffer doesn't have an init list ctor

std::vector<int> v(10); // makes a vector of 10 default initialized ints
std::vector<int> v{10}; // makes a vector with 1 int element initialized to 10
```

This is widely regarded as a pretty significant failing of C++ when they added brace initialization, but it is what it is, and life goes on!

File: cpp/include/cudf/detail/gather.cuh (hunk 2 above)
Comment (jrhemstad) [reply cpp/include/cudf/detail/gather.cuh:248]: Yeah, that about matches what I expected. 

> Do you see a downside to this change?

Not really, just more verbose.

**Comment by jrhemstad** [comment]
> I noticed a few days ago that `device_buffer` holds on to the `mr` value and has `resize()`. This means that the passed `mr` can potentially be used even if the initial buffer size is zero and no allocation happens at construction time. Given that it's somewhat easier to use `mr` consistently on all returned buffers (empty or not), and there's no overhead from passing the correct `mr`, I went for it.

That's kind of what I figured. I think you are correct to make this change for safety's sake.

File: cpp/include/cudf/detail/reduction.cuh (hunk 1 above)
Comment (vuule) [reply cpp/include/cudf/detail/reduction.cuh:103]: Yeah, I'm aware of the issues with {}, so (misremembering Conor's advice) I thought we just want to avoid it, unless init list is actually used.

File: cpp/src/groupby/groupby.cu [hunk 3]
```diff
@@ -146,7 +146,8 @@ groupby::groups groupby::get_groups(table_view values, rmm::mr::device_memory_re
 
   std::unique_ptr<table> grouped_values{nullptr};
   if (values.num_columns()) {
-    grouped_values = cudf::experimental::detail::gather(values, helper().key_sort_order());
+    grouped_values = cudf::experimental::detail::gather(
+      values, helper().key_sort_order(), false, false, false, mr);
```
Comment (codereport) [inline cpp/src/groupby/groupby.cu:150]: This is kind of API that advocate trying to avoid using scoped enumerators.
```cpp
false, false, false
```
Looks like this is pre-existing - I can set up an issue to address this in a follow PR. 
```cpp
std::unique_ptr<table> gather(table_view const& source_table,
                              column_view const& gather_map,
                              bool check_bounds,
                              bool ignore_out_of_bounds,
                              bool allow_negative_indices,
                              rmm::mr::device_memory_resource* mr,
                              cudaStream_t stream)
```

File: cpp/src/column/column_factories.cpp [hunk 4]
```diff
@@ -128,20 +128,23 @@ struct column_from_scalar_dispatch {
...
              cudaStream_t stream) const
   {
-    auto null_mask = create_null_mask(size, mask_state::ALL_NULL, stream, mr);
     if (!value.is_valid())
-      return std::make_unique<column>(
-        value.type(), size, rmm::device_buffer{}, std::move(null_mask), size);
+      return std::make_unique<column>(value.type(),
+                                      size,
+                                      rmm::device_buffer{0, stream, mr},
+                                      create_null_mask(size, mask_state::ALL_NULL, stream, mr),
```
Comment (codereport) [inline cpp/src/column/column_factories.cpp:135]: Love this change! :) even though just cosmetic.

File: cpp/src/column/column_factories.cpp [hunk 5]
```diff
@@ -128,20 +128,23 @@ struct column_from_scalar_dispatch {
...
+                                      size);
 
     // Create a strings column_view with all nulls and no children.
     // Since we are setting every row to the scalar, the fill() never needs to access
     // any of the children in the strings column which would otherwise cause an exception.
+    auto null_mask = create_null_mask(size, mask_state::ALL_NULL, stream);
```
Comment (codereport) [inline cpp/src/column/column_factories.cpp:141]: Consider making the same change here as you did above? aka removing the temporary

File: cpp/src/merge/merge.cu [hunk 6]
```diff
@@ -457,8 +457,17 @@ table_ptr_type merge(std::vector<table_view> const& tables_to_merge,
...
+
+    // Only use mr for the output table
+    auto const& new_tbl_rm = merge_queue.empty() ? mr : rmm::mr::get_default_resource();
+    auto merged_table      = merge(left_table.view,
+                              right_table.view,
+                              key_cols,
+                              column_order,
+                              null_precedence,
+                              new_tbl_rm,
+                              stream);
```
Comment (codereport) [inline cpp/src/merge/merge.cu:469]: Interesting - this is clearly a `clang-format` bug - might be fixed in a clang version > 8. If should be be aligning the arguments **after** it aligns the equal sign (instead of the reverse which it is clearly doing). It should be:
```cpp
    auto const& new_tbl_rm = merge_queue.empty() ? mr : rmm::mr::get_default_resource();
    auto merged_table      = merge(left_table.view,
                                   right_table.view,
                                   key_cols,
                                   column_order,
                                   null_precedence,
                                   new_tbl_rm,
                                   stream);
```

File: cpp/src/strings/combine.cu [hunk 7]
```diff
@@ -76,8 +76,8 @@ std::unique_ptr<column> concatenate(
     },
     stream,
     mr);
-  rmm::device_buffer null_mask = valid_mask.first;
-  auto null_count              = valid_mask.second;
+  auto& null_mask       = valid_mask.first;
```
Comment (codereport) [inline cpp/src/strings/combine.cu:79]: :+1:

File: cpp/src/groupby/groupby.cu (hunk 3 above)
Comment (vuule) [reply cpp/src/groupby/groupby.cu:150]: Yeah, it felt icky adding these parameters to the call.

File: cpp/src/column/column_factories.cpp (hunk 5 above)
Comment (vuule) [reply cpp/src/column/column_factories.cpp:141]: I don't think I can do it here, the mask buffer is used in multiple statements. The ownership is not transferred in the `column_view sc` creation, so I have to keep `null_mask` alive until `strings::detail::fill` is done.

**Comment by harrism** [comment]
Does this provide a fix for #5067? If so, use the syntax "Fixes #<ISSUE #>" in your description.

File: cpp/src/column/column_factories.cpp (hunk 5 above)
Comment (codereport) [reply cpp/src/column/column_factories.cpp:141]: Mmm i see. Makes sense
